### PR TITLE
[ETCM-78 added checkpointing JRC

### PR DESCRIPTION
**Pull request by pslaski**
# Description

Ported Checkpointing JRC

# Important Changes Introduced

- new checkpointing RPC API
- new endpoint - checkpointing_getLatestBlock
- new endpoint checkpointing_pushCheckpoint

# Testing

use insomnia endpoints

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/jsonrpc/JsonRpcController.scala [hunk 1]
```diff
@@ -69,7 +72,7 @@ object JsonRpcController {
         override val apis: Seq[String] = {
           val providedApis = rpcConfig.getString("apis").split(",").map(_.trim.toLowerCase)
           val invalidApis =
-            providedApis.diff(List("web3", "eth", "net", "personal", "daedalus", "test", "iele", "debug", "qa"))
+            providedApis.diff(List("web3", "eth", "net", "personal", "daedalus", "test", "iele", "debug", "qa", "checkpointing"))
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/jsonrpc/JsonRpcController.scala:75]: Minor: Should we use the variables from the Apis object here instead?

File: src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingJRCSpec.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    val response = jsonRpcController.handleRequest(pushCheckpointRequest).futureValue
+    response should haveResult(expectedResult)
+  }
+
+  object Req {
+    val block = Fixtures.Blocks.ValidBlock.block
+    val signatures = List(
+      "cca1f7599f24c2679a6da2797f39b44b4d863c3fcd5cd39e38e9f3d38f1934db53668796ece8cdbf44b1a5b78326cc8a0c48c030ae26cac78ab00a79653b20781b",
+      "aab5f8761636139990cd406476392ef1f9b7f3a497155be17e5c9fe558afa1e1d1f2f470b659ac772e56a9e6f0b03b15f335cc2db8842af1c7130112c88c2af41c"
```
Comment (ntallar) [inline src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingJRCSpec.scala:64]: Should we maybe produce the signatures here to not have them hardcoded?

File: src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingJRCSpec.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    with Matchers
+    with MockFactory
+    with ScalaFutures
+    with NormalPatience
+    with JRCMatchers
+    with JsonMethodsImplicits {
+
+  import Req._
+
+  "CheckpointingJRC" should "getLatestBlock" in new TestSetup {
```
Comment (ntallar) [inline src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingJRCSpec.scala:28]: Should we maybe add tests for the non happy cases as when the signatures can't be decoded?

File: src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+class CheckpointingServiceSpec
+  extends TestKit(ActorSystem("CheckpointingServiceSpec_System"))
+    with AnyFlatSpecLike
+    with MockFactory
+    with ScalaFutures
+    with NormalPatience
+    with ScalaCheckPropertyChecks
+    with Matchers {
+
+  "CheckpointService" should "get latest block (at a correct checkpointing interval) from Blockchain" in new TestSetup {
```
Comment (ntallar) [inline src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingServiceSpec.scala:25]: Should we maybe add a test for the case a reorg happened?

**Review by ntallar — CHANGES_REQUESTED** [review]
Only minor comments and LGTM!

File: src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingServiceSpec.scala (hunk 4 above)
Comment (rtkaczyk) [reply src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingServiceSpec.scala:25]: Should be doable by setting expectations on blockchain accordingly, right?

File: src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingJRCSpec.scala [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    val expectedResult = JObject(
+      "hash" -> JString("0x" + ByteStringUtils.hash2string(block.hash)),
+      "number" -> JInt(block.number)
+    )
+
+    val response = jsonRpcController.handleRequest(request).futureValue
+    response should haveResult(expectedResult)
+  }
+
+  it should "getLatestBlock in case of blockchain re-org" in new TestSetup {
```
Comment (ntallar) [inline src/test/scala/io/iohk/ethereum/jsonrpc/CheckpointingJRCSpec.scala:50]: Isn't this more a test for the `CheckpointingServiceSpec`?

**Review by ntallar — CHANGES_REQUESTED** [review]
Apart from a minor comment, LGTM!

**Review by rtkaczyk — COMMENTED** [review]
LGTM, except for Nico's comment about the test

**Review by rtkaczyk — APPROVED** [review]

**Review by ntallar — APPROVED** [review]
LGTM!
```
 .\\            //.
. \ \          / /.
.\  ,\     /` /,.-
 -.   \  /'/ /  .
 ` -   `-'  \  -
   '.       /.\`
      -    .-
      :`//.'
      .`.'
      .' BP 
```

**Pull request merged**